### ISSUE TITLE
Initialization Data Encountered Algorithm: 'encrypted' event should not be fired if there is no action to be taken

### ISSUE THREAD

**Issue by greg80303** [w3c/encrypted-media#121]
The current EME spec indicates that the user agent will send the `MediaEncryptedEvent` every time it encounters InitializationData in the media data.  For adaptive bitrate protocols like MPEG-DASH this may cause the delivery of the `MediaEncryptedEvent` upon every Representation switch.  The Javascript application must then decide whether or not to call `MediaKeySession.generateRequest()` with that InitializationData based on a very limited set of information.  The CDM, being the only entity that can decipher the contents of the InitializationData, should manage this process without any intelligence on the part of the application.  A couple of possible solutions:

* All instances of InitializationData are passed to the CDM automatically by the user agent upon detection in the media data.  The `MediaEncryptedEvent` will be sent ONLY when a transaction with the license server is required.
* Continue to deliver `MediaEncryptedEvent` each time InitializationData is encountered.  The application will continue to call `MediaKeySession.generateRequest()` for each occurrence, but modify the return value of the function to be `Promise<boolean>`, where a value of true indicates that a license request will be generated, and a value of false indicates that no license request is necessary.

Additionally, in the case of “broadcast” delivery of content, where the client device does not always have an upstream data path, the content distributor may place InitializationData in the media data that contains encrypted keys.  These encrypted keys can be decrypted by the CDM using a “root” license that was obtained by other methods.  This use-case relies on a solution to the previously mentioned problem in which the CDM automatically “handles” InitializationData for which no action is required by the Javascript application.  However, in this case, you would expect `keystatuseschange` events to be delivered to the application when the CDM processes and installs these “in-band licenses”.

**Comment by ddorwin** [comment]
The group tried to address this before [bug 25268](https://www.w3.org/Bugs/Public/show_bug.cgi?id=25268) but were unable to come up with a good solution. Suggestions welcome, but I think that will likely need to be addressed in a future version.

Regarding the first possible solution, the CDM is _intentionally_ not involved in the processing (demuxing) of media data. Specifically, the CDM is not used in `HTMLMediaElement`'s [Initialization Data Encountered](https://w3c.github.io/encrypted-media/#initdata-encountered) algorithm.

The last paragraph seems like one or more separate feature requests. I have questions, but that should probably be discussed separately.

**Comment by tinskip** [comment]
The MediaEncryptedEvent is generated by the user agent, and not necessarily from within the EME implementation, but more likely from the media stack, which has no knowledge of media keys.

I'd like to propose a third alternative: adding a new method to MediaKeys interface, which can be used to query the underlying CDM implementation whether a new MediaKeySession (plus associated key request) is required.

boolean haveSession(DOMString initDataType, BufferSource initData);

Upon handling of a MediaEncryptedEvent, the application would just call MediaKeys.haveSession in order to determine whether it should create a new session and call MediaKeySession.generateRequest(). This solution has the advantages of being able to keep the knowledge of keys restricted to the EME implementation and CDM, not having to create useless MediaKeySession's which have to be abandoned after generateRequest fails, and is completely backward compatible with the existing interface and any deployed applications.

**Comment by joeyparrish** [comment]
The proposal from @tinskip makes sense to me as an author.  If the new method isn't available on a particular browser, a simple polyfill could just return false (or Promise.resolve(false) if it ends up async), which would cause my app to continue with current behavior and potentially have redundant sessions.  His proposal is also backward compatible with existing apps for the same reason.

**Comment by greg80303** [comment]
I also like the proposal from @tinskip better than my original suggestion.  Much cleaner to not create key sessions for every InitData and have to tear them down if the data is redundant or unnecessary.

**Comment by jdsmith3000** [comment]
I'm interested in the proposed MediaKeys method and think it deserves discussion in the Media Task Force.

It's proposed that this method be used to both limit duplicate MediaKeySessions (by checking with the CDM directly to see if a key is already available for the initData) and to support in-media encrypted content keys that are decrypted by a root key obtained separately.  This second goal would raise questions about how existing algorithms apply to these in-media keys (obtained without a MediaKeySession) and what concept of session should be applied to them, if any.

@greg80303, @joeyparrish, @tinskip:  How similar is this to the question raised in #53?  And what importance do you place on solving it for v1?

**Comment by greg80303** [comment]
@jdsmith3000 For our interests, the first part of the solution is of higher priority (eliminating duplicate key sessions).  We would love for this to be fixed in v1, because we currently don't have a working solution for the test content that led to the discovery of this problem.

**Comment by ddorwin** [comment]
@greg80303, can you provide more information about the content and application? Specifically, why other solutions, such as obtaining initData from a manifest or ignoring subsequent `encrypted` events, won't work.

While @tinskip's proposal sounds promising, I think this is something we should incubate and get implementation experience with before etching it in stone. The relation to in-media keys also suggests we  need to to consider more use cases and iterate, something better suited for vNext.

**Comment by jdsmith3000** [comment]
I agree that we need to be careful about adding use cases at this point in time, but we need to be sure we understand the importance of the request as well.  I'm not sure currently that we have a good read on that, and am hearing through other channels that the inband key use case is very important.  That's why I asked above.

**Comment by greg80303** [comment]
@ddorwin [Here](http://b028.wpc.azureedge.net/80B028/Samples/859b335f-eebc-4ff1-ac03-98e9b614e2c1/0cc164eb-193e-48cb-bc6b-041cc28307e4.ism/Manifest(format=mpd-time-csf)) is an example piece of MPEG-DASH content that caused me to generate the discussion around this problem.  This content utilizes the CommonEncryption scheme and initialization data format.

In this content, the DASH manifest contains initData and a CommonEncryption ContentProtection element with a `default_keyID` attribute.  The media also contains initData in the initialization segments.  However, the initData in the media (v0 PSSH) is not a binary match to the initData in the DASH manifest (PlayReady PSSH initData allows for some *padding* bytes which supports extension of the PSSH later without changing the ISOBMFF box size).  With a v0 PSSH, there is no way for my application to know if the initData in the media is *unique* and prevent the creation of duplicate key sessions and license requests.

In the end, maybe the problem is not with EME, but with the initData registry.  Maybe there should be a requirement among initData formats for an *equality* test to avoid this problem.  In many systems, like MPEG-DASH, initData can be delivered out-of-band as well as in-band.  If the JavaScript application is going to be responsible for comparing the opaque, DRM-specific, initData blobs delivered via these two mechanisms, there should be a well-defined, DRM-independent way to do so.  CommonEncryption v1 KeyIDs and DASH ContentProtection default_keyID fields could be used to satisfy this requirement, but it would require some clarification in other specifications.

**Comment by mwatson2** [comment]
Greg, could you explain in more detail this use-case where the client application does not know whether the manifest-provide or content-provided initData has precedence ? I think we have assumed to date that the application knows which source it will use for initData.

**Comment by greg80303** [comment]
This is not a matter of precendence.  In our case, the manifest initData always takes precedence.  So, I create a key session and retrieve a license based on the manifest initData. Then, then we start placing media into the source buffers and we get an `encrypted` event with initData.  How does the application determine whether it should initiate a new key session or if the initData passed in the encrypted event is just a "duplicate" of something that was already handled?  Additionally, since the initData will be found in every initialization segment (for most DASH content), every time we switch Representations, we'll get a new `encrypted

**Comment by greg80303** [comment]
@mwatson2 Again, maybe the problem here is more specific to MPEG-DASH/CommonEncryption and needs to be handled in other specs.  But since the CDM is something that *could* determine whether or not a new key session needs to be initiated, I wanted to check with this group to see if EME is the right place to address the problem.

**Comment by mwatson2** [comment]
EME is likely the right place to address the problem, but we need to fully understand the problem first.

What is the use-case where you need to mix manifest-provided and in-band initData ? It's not clear to me why, if you can get some initData from the manifest you wouldn't get all initData that way ?

Secondly, when you have multiple streams and stream-switching, what is the use-case where the application does not know which streams share keys and which have distinct keys ? Why would not not obtain all the needed keys for all the streams at the start of the session ?

**Comment by tinskip** [comment]
I believe there is a point which is being overlooked: the application
should not be responsible for implementing DRM-system specific logic.  We
can require that the initData contain key IDs in a normalized format,
accessible to the application, require that CDMs report key statuses for
all keys, and have applications implement the logic which decides whether a
new session + key request is necessary by comparing keys IDs in the PSSH
against the reported key IDs in the keyStatus. But who's to say that a
yet-unknown DRM system is not going to need to start a new session based on
some criterion other than key availability?

Also, it seems much cleaner to me to have this logic implemented in the CDM
than to add complexity to every player application.


On Wed, Nov 18, 2015 at 8:44 AM, mwatson2 <notifications@github.com> wrote:

> EME is likely the right place to address the problem, but we need to fully
> understand the problem first.
>
> What is the use-case where you need to mix manifest-provided and in-band
> initData ? It's not clear to me why, if you can get some initData from the
> manifest you wouldn't get all initData that way ?
>
> Secondly, when you have multiple streams and stream-switching, what is the
> use-case where the application does not know which streams share keys and
> which have distinct keys ? Why would not not obtain all the needed keys for
> all the streams at the start of the session ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/w3c/encrypted-media/issues/121#issuecomment-157773386>
> .
>



-- 
Thomás Inskip | Software Engineer | tinskip@google.com | +1 650 241 9697

**Comment by mwatson2** [comment]
@tinskip CDMs can initiate message exchanges within the context of existing sessions at any time. The only purpose of "sessions" is to provide multiple message exchange contexts to enable applications to have multiple such exchanges proceed in parallel (otherwise we would just have all the methods / messages on the MediaKeys object and not have sessions at all).

So there is no reason that a CDM would need to trigger a new session: it can always use an existing one.

My questions are really trying aimed at better understanding the use-case, because it feels like there are a lot of unstated assumptions. Particularly there are things which suggest you assume the application has less knowledge of the nature of the content encryption that I would expect,

**Comment by greg80303** [comment]
@mwatson2 I agree with you in that a player within a vertically integrated ecosystem would have some expectations of the content that it will be presenting above and beyond what is stated in the specifications.  Unfortunately, I am working on a generic player library that relies on MPEG-DASH, DASH-IF IOP, and ISO CommonEncryption (in addition to MSE/EME) for its "requirements" and not much more.  The scenario I describe above (initData in manifest AND in media) is not prohibited by any of these specifications, so I'm trying to handle situation in my code if/when it arises.  The test content I referenced above exhibits these symptoms, but I have no idea if anyone is actually planning to deploy content in this fashion and what the business use-cases would be for doing so.

**Comment by joeyparrish** [comment]
@greg80303, my understanding is that when a DASH manifest specifies a PSSH, it takes precedence over the PSSH in the content, such that you simply don't need to listen for the encrypted event at all.

**Comment by mwatson2** [comment]
@greg80303 The ability to implement generic player components is important, so it would be good to better understand what requirements this might place on the EME API.

A "generic player" basically means that what we have previously considered as a monolithic "application" (Client JS communicating with some server components) is functionally split into a service-specific part and a generic part. We need to understand this functional split and any constraints on that split to understand the requirements on EME.

For example, I suspect that one assumption in your use-case is that the interface between service-specific component and generic player is a DASH manifest. So, for example, that interface does not provide support for the service-specific part of the application to inform the generic part about the arrangement of streams and keys etc. and thus you are expecting to be able to work this out from the manifest and streams alone. Is this correct ?

**Comment by greg80303** [comment]
@mwatson2 Mark, I think your stated assumption is correct.  I don't think the DASH manifest has enough information for me to be able to determine the disposition of each `encrypted` event.  As a player application, the manifest is all I have at my disposal.  Even though I do have access to the streams as I push them into source buffers, it is not reasonable to expect the app to parse them looking for initData (hence, the `encrypted` event).

**Comment by cpearce-mozilla** [comment]
Is it the case that the same key IDs can be encoded into multiple different initData? That is, you can't at the application layer just keep a map of initData you've passed to which MediaKeySession's generateRequest() and use that to figure out whether you need already have a MediaKeySession for new initData that you encounter?

Assuming yes, I think @tinskip's proposal would be better as:

promise<bool> haveSession(DOMString initDataType, BufferSource initData)

If the implementation has to call into the CDM, then it *must* be asynchronous.

Is it necessary/helpful to identify the MediaKeySession associated with the initData? An alternative would be:

promise<MediaKeySession> haveSession(DOMString initDataType, BufferSource initData)

Where the CDM identifies to the userAgent the MediaKeySession instance, which is returned to JS.
